lib/vrl/stdlib/src/parse_csv.rs

```diff
@@ -0,0 +1,127 @@
+use bytes::Buf;
+use csv::ReaderBuilder;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseCsv;
+
+impl Function for ParseCsv {
+    fn identifier(&self) -> &'static str {
+        "parse_csv"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "parse a single CSV formatted row",
+            source: r#"parse_csv!(s'foo,bar,"foo "", bar"')"#,
+            result: Ok(r#"["foo", "bar", "foo \", bar"]"#),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        let delimiter = arguments.optional("delimiter").unwrap_or(expr!(","));
+        Ok(Box::new(ParseCsvFn { value, delimiter }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "delimiter",
+                kind: kind::BYTES,
+                required: false,
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseCsvFn {
+    value: Box<dyn Expression>,
+    delimiter: Box<dyn Expression>,
+}
+
+impl Expression for ParseCsvFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let csv_string = self.value.resolve(ctx)?.try_bytes()?;
+        let delimiter = self.delimiter.resolve(ctx)?.try_bytes()?;
+        if delimiter.len() != 1 {
+            return Err("delimiter must be a single character".into());
+        }
+        let delimiter = delimiter[0];
+
+        let reader = ReaderBuilder::new()
+            .has_headers(false)
+            .delimiter(delimiter)
+            .from_reader(csv_string.bytes());
+
+        let result = reader
+            .into_byte_records()
+            .next()
+            .transpose()
+            .map_err(|err| format!("invalid csv record: {}", err))? // shouldn't really happen
+            .map(|record| record.iter().map(Into::into).collect::<Vec<Value>>())
```

Comment:
nit

```suggestion
            .map(|record| record.iter().map(Into::into).collect::<Vec<_>>())
```

When collecting an iterator, you often have to help the compiler to make it understand you want to collect in a `Vec`, but it can figure out the rest, so you can use `_` to reduce the amount of explicit types needed.

Reply:
nice 👍

Reply:
ah, but then it doesn't work because it can't infer the type for `map`. 
```
.map(|record| record.iter().map(Into::into).collect::<Vec<_>>())
   |                                         ^^^ cannot infer type for type parameter `B` declared on the associated function `map`
   |
help: consider specifying the type arguments in the method call
   |
68 |             .map(|record| record.iter().map::<B, F>(Into::into).collect::<Vec<_>>())
```
So I guess we still need to leave the type here(or specify it for `map`).